src/concrete/ml/quantization/quantized_module_passes.py

```diff
@@ -0,0 +1,300 @@
+                integer_node_input_quant: Optional[QuantizedOp] = None
+
+                while back_node_output != prev_compatible_node_output:
+                    assert back_node is not None
+                    nodes_in_path.append(back_node)
+                    assert_true(
+                        back_node in predecessors,
+                        "Power of Two adapter: Error during graph traversal",
+                    )
+                    # If multiple ops produced this node, the pattern is not matched
+
+                    if len(predecessors[back_node]) > 1:
+                        break
+
+                    back_node, back_node_output = predecessors[back_node][0]
+
+                    # Reached the previous integer node
+                    if back_node_output == prev_compatible_node_output:
+                        # The Gemm/Conv op that produces this integer node is the one
+                        # onto which we apply the roundPBS optimization
+                        nodes_in_path.append(back_node)
+                        list_pred_of_path = predecessors[back_node]
+                        if len(list_pred_of_path) == 1:
+                            integer_node_input_quant = list_pred_of_path[0][0]
+
+                assert isinstance(node_op, QuantizedMixingOp)
+                if self.match_path_pattern(predecessors, nodes_in_path, integer_node_input_quant):
+                    # If rounding was manually set (usually globally for all layers)
+                    # the do not override the requested number of rounding bits
+                    # but keep statistics for testing purposes
+                    path_start_node = nodes_in_path[-1]
+                    assert isinstance(path_start_node, QuantizedMixingOp)
+                    if path_start_node.rounding_threshold_bits is not None:
+                        self._num_ignored_valid_patterns += 1
+                    else:
+                        valid_paths[path_start_node] = (nodes_in_path, integer_node_input_quant)
+        return valid_paths
+
+    def process_patterns(self, valid_paths: PatternDict) -> PatternDict:
+        """Configure the rounding bits of roundPBS for the optimizable operations.
+
+        Args:
+            valid_paths (PatternDict): list of optimizable patterns
+
+        Returns:
+            result (PatternDict): list of patterns actually optimized with roundPBS
+        """
+
+        def integer_log2(value: float) -> Tuple[int, bool]:
+            """Compute the log2 of the value and tests if its an integer.
+
+            Args:
+                value (float): the value for which to take the log2
+
+            Returns:
+                result: The integer log2 and a bool indicating whether
+                    the input value was an integer power of two
+            """
+            log2_value = int(numpy.rint(numpy.log2(value)))
+            if numpy.isclose(numpy.power(2.0, log2_value), value, atol=0.01):
```

Comment:
is it not enough t just check that `numpy.rint(numpy.log2(value)) == numpy.log2(value)` ? or am I missing something ? 

if not, then maybe make "atol" a parameter for `integer_log2` (or at least say why it was set to 0.01)

Reply:
the best practice is to never compare floats using equality. But you're right about the 0.01, it's quite arbitrary and not a good value for low powers of two (it's too big when the power is say -7). I'll use rtol instead